instrumentor/controllers/startlangdetection/source_controller.go

```diff
@@ -30,54 +31,100 @@ func (r *SourceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
+		}
 
-			if source.Labels == nil {
-				source.Labels = make(map[string]string)
-			}
-			source.Labels[consts.WorkloadNameLabel] = source.Spec.Workload.Name
-			source.Labels[consts.WorkloadNamespaceLabel] = source.Spec.Workload.Namespace
-			source.Labels[consts.WorkloadKindLabel] = string(source.Spec.Workload.Kind)
+		source.Labels[consts.WorkloadNameLabel] = source.Spec.Workload.Name
+		source.Labels[consts.WorkloadNamespaceLabel] = source.Spec.Workload.Namespace
+		source.Labels[consts.WorkloadKindLabel] = string(source.Spec.Workload.Kind)
 
-			if err := r.Update(ctx, source); err != nil {
-				return k8sutils.K8SUpdateErrorHandler(err)
-			}
+		if err := r.Update(ctx, source); err != nil {
+			return k8sutils.K8SUpdateErrorHandler(err)
+		}
 
-			err = requestOdigletsToCalculateRuntimeDetails(ctx, r.Client, instConfigName, req.Namespace, obj, r.Scheme)
+		// pre-process existing Sources for specific workloads so we don't have to make a bunch of API calls
+		// This is used to check if a workload already has an explicit Source, so we don't overwrite its InstrumentationConfig
+		sourceList := v1alpha1.SourceList{}
+		err := r.Client.List(ctx, &sourceList, client.InNamespace(source.Spec.Workload.Name))
+		if err != nil {
 			return ctrl.Result{}, err
 		}
-	} else {
-		// Source is being deleted
-		if controllerutil.ContainsFinalizer(source, consts.SourceFinalizer) {
-			// Remove the finalizer first, because if the InstrumentationConfig is not found we
-			// will deadlock on the finalizer never getting removed.
-			// On the other hand, this could end up deleting a Source with an orphaned InstrumentationConfig.
-			controllerutil.RemoveFinalizer(source, consts.SourceFinalizer)
-			if err := r.Update(ctx, source); err != nil {
+		namespaceKindSources := make(map[workload.WorkloadKind]map[string]struct{})
+		for _, source := range sourceList.Items {
+			if _, exists := namespaceKindSources[source.Spec.Workload.Kind]; !exists {
+				namespaceKindSources[source.Spec.Workload.Kind] = make(map[string]struct{})
+			}
+			// ex: map["Deployment"]["my-app"] = ...
+			namespaceKindSources[source.Spec.Workload.Kind][source.Spec.Workload.Name] = struct{}{}
+		}
+
+		if source.Spec.Workload.Kind == "Namespace" {
+			var deps appsv1.DeploymentList
+			err = r.Client.List(ctx, &deps, client.InNamespace(source.Spec.Workload.Name))
+			if client.IgnoreNotFound(err) != nil {
+				logger.Error(err, "error fetching deployments")
+				return ctrl.Result{}, err
+			}
+
+			for _, dep := range deps.Items {
+				if _, exists := namespaceKindSources[workload.WorkloadKindDeployment][dep.Name]; !exists {
+					request := ctrl.Request{NamespacedName: client.ObjectKey{Name: dep.Name, Namespace: dep.Namespace}}
+					_, err = reconcileWorkload(ctx, r.Client, workload.WorkloadKindDeployment, request, r.Scheme)
+					if err != nil {
+						logger.Error(err, "error requesting runtime details from odiglets", "name", dep.Name, "namespace", dep.Namespace)
```

Comment:
Since we have this exact message 3 times, I suggest also recording the workload kind here, so it's easier to track the relevant place if it gets printed